### DISCUSSION TITLE
Having trouble getting content_for to work for passing data to layout

### DISCUSSION THREAD

**Discussion by jarednorman** [yippee-fun/phlex#555] [Help]
I'm working on a small Rails application, trying to convert everything over to Phlex. It's so far gone smoothly, except I can't get `content_for`/`yield` to work as described [here](https://www.phlex.fun/rails/layouts/).

I've tried a couple of approaches, but none worked.

Within my `ApplicationLayout`'s `template` method, I have:

```ruby
render SiteFooterComponent.new(
  copyright_year: yield(:copyright_year)
)
```

Regardless of which of the following approaches I take, `yield(:copyright_year)` always returns `nil`.

1. I've tried calling `content_for` from the controller. However, from browsing Stack Overflow, there's some question about whether that works.

   ```ruby
   # Inside PostsController#show
   post = Post.find_by!(slug: params[:path])

   helpers.content_for(:copyright_year) { post.published_at&.year }

   render Posts::ShowView.new(post: post)
   ```

2. I've also tried calling `content_for` inside `Posts::ShowView#template`, as follows. The result is the same.

   ```ruby
   content_for(:copyright_year) {
     # I tried letting the block return the value and using the text and plain helpers. No difference.
     text @post.published_at&.year
   }

   render BlogHeaderComponent.new(post: @post)

   render "markdown_content", markdown_content: @post.body_html
   ```

   I tried both by including `Phlex::Rails::Helpers::ContentFor` in the view and by calling `helpers.content_for`. I didn't expect that to make a difference, and it didn't.

Do you have any thoughts on what I'm doing wrong here?

**Comment by joeldrapper** [comment]
Hey @jarednorman, I will take a proper look at this tomorrow. In the meantime, could you try `capture { yield(:foo) }` to see if that works?

**Reply by jarednorman** [reply]
That works, though I'm curious as to why.

**Reply by jarednorman** [reply]
It really would be lovely to find a way to allow the layouts to be parameterized. I'm just looking at why we're limited there.

Edit: Ah, the `render` methods collide... awkward.

**Comment by jarednorman** [comment]
I am setting this aside as a separate thread, as it's directly related but different from the exact issue I experienced.

I was able to parameterize the layouts using the following code. Of course, this is gross, and I'm sure a better solution exists. However, parameterized layouts are essential for my use case, so I want this snippet to be available for anyone who wants it, even if phlex-rails doesn't directly support it.

I put the following into my `ApplicationLayout` class.

```ruby
class Wrapper
  def initialize(klass:, args:, kwargs:)
    @klass = klass
    @instance = klass.new(*args, **kwargs)
  end

  def render(view, _locals, &block)
    @instance.call(view_context: view) do |yielded|
      case yielded
      when Symbol
        output = view.view_flow.get(yielded)
      else
        output = yield
      end

      case output
      when ActiveSupport::SafeBuffer
        @instance.unsafe_raw output
      end

      nil
    end
  end

  def identifier
    @klass.identifier
  end

  def virtual_path
    @klass.virtual_path
  end
end

def self.with(*args, **kwargs)
  Wrapper.new(klass: self, args: args, kwargs: kwargs)
end
```

This allows at the controller level for specifying the layout like in the docs (`layout -> { ApplicationLayout }`) and when you need parameters, using the `with` method: `layout -> { ApplicationLayout.with(hide_header: true) }`.
